src/cryptonote_core/blockchain.cpp

```diff
@@ -3039,12 +3042,10 @@ bool Blockchain::check_tx_outputs(const transaction& tx, tx_verification_context
   // from v4, forbid invalid pubkeys
   if (hf_version >= 4) {
     for (const auto &o: tx.vout) {
-      if (o.target.type() == typeid(txout_to_key)) {
-        const txout_to_key& out_to_key = boost::get<txout_to_key>(o.target);
-        if (!crypto::check_key(out_to_key.key)) {
-          tvc.m_invalid_output = true;
-          return false;
-        }
+      crypto::public_key output_public_key;
+      if (get_output_public_key(o, output_public_key) && !crypto::check_key(output_public_key)) {
+        tvc.m_invalid_output = true;
+        return false;
```

Comment:
1. Should this code be checking if the types of `o.target` are all valid? What happens if there is an unknown target type?
2. I think it would be cleaner to separate the `&&` operands here. It is kind of sketchy to have one part of an expression depend on another part of the expression (even if logically valid).

Reply:
>1. Should this code be checking if the types of o.target are all valid? What happens if there is an unknown target type?

My thinking was maintain the logic of what it's currently doing (check for valid public keys), and enforce expected type elsewhere (as is currently done in `check_outs_valid` and when adding an output to LMDB). I'll dig deeper to see if it's definitively ok to enforce the exact type here.

> 2. I think it would be cleaner to separate the && operands here.

Fair

Reply:
Ok, check out the latest commit [9d9d2c](https://github.com/monero-project/monero/pull/8061/commits/9d9d2c613caeeba1076e0a34d12b50a4bc161f53). It's now enforcing exact expected types on the outputs here and in `check_output_types` (I checked that this should be ok pretty thoroughly).

To answer this question:

> What happens if there is an unknown target type?

Assuming you meant a type not defined in the boost variant, it won't parse. Current nodes won't be able to parse blocks or transaction blobs that have view tags, since to those nodes, the outputs are of an unknown target type